### ISSUE TITLE
Webdav sync is broken since 2.12.15

### ISSUE THREAD

**Issue by romu70** [laurent22/joplin#8788]
## Environment

Joplin version: 2.12.15
Platform: macOS Ventura 13.5.1 Arm64
OS specifics: 

## Steps to reproduce

I sync my iPad, my iPhone and my Mac on my Synology nas through Webdav. So of course, with the same settings.

1. Open Joplin on the Mac
2. Go to settings -> Synchronisation
3. Check the sync settings are the same than on my other devices (it's not a fresh install, I've been using Joplin for a long time).
4. Click on "Check synchronisation configuration"

And I get:

> Error. Please check that URL, username, password, etc. are correct and that the sync target is accessible. The reported error was:
> request to https://---my webdav server url---/home/notes/ failed, reason: connect ECONNREFUSED ---my home ipv6 address---:443 (Code ECONNREFUSED)

## Describe what you expected to happen

On my other devices, I get this:

> Success! Synchronisation configuration appears to be correct.

**Comment by romu70** [comment]
Oups, I forgot to attach the log trace.

![Screenshot 2023-09-06 at 08 26 03](https://github.com/laurent22/joplin/assets/1898319/4d316f30-a7ae-4954-9b25-15d617310a18)

**Comment by PaulTate** [comment]
WebDav sync also appears broken on Windows.  I am using an https url.

![image](https://github.com/laurent22/joplin/assets/34127627/f4c7fb1b-6aaf-4b64-8987-c07086b31fb6)

Joplin 2.12.15 (prod, win32)

Client ID: 4cfbbe01b63a4cb59e341522d9a64a9c
Sync Version: 3
Profile Version: 43
Keychain Supported: Yes

Revision: 2d26332

Downgrading to 2.9.17 works.

**Comment by romu70** [comment]
The last prod macOS version is 2.10.19, and as I already ran the 2.12.x, I can't downgrade. I'm stuck with non synchronising files.

**Comment by JackGruber** [comment]
WebDav Sync works without problems for me on Windows 10 with Joplin 2.12.15 to my Synology WebDav service via https.

**Comment by romu70** [comment]
Same issue on 2.15.16.

**Comment by NicholasYamamoto** [comment]
Started seeing this issue too on my iOS and MacOS Ventura device. Confirmed that I am able to reach my WebDAV just fine, but Joplin will not sync with it. I am also being shown `Success! Synchronization configuration appears to be correct.` on both of my devices when checking the sync configuration, despite it not actually working.

**Comment by romu70** [comment]
Same with 2.12.18. And as I don’t have any workaround and need my notes, it’s the end for me with Joplin. Too bad.

**Comment by jogi4github** [comment]
I was able to solve my ECONNREFUSED sync issue on my macbook by disabling ipv6.

```
network -listallnetworkservices
networksetup -setv6off <networkservice>
```

**Comment by V1ck3s** [comment]
I have the same problem:
`
FetchError: request to http://myjoplin/api/sessions failed, reason: connect ECONNREFUSED 2a1:ipv6::1:80
`

I tested several versions, version 2.12.5 works and it is from the next version, 2.12.7 that it starts to no longer work.

The changelog for the 2.12.7 is : 

> Improved: Updated packages css-loader (v6.8.1)
    Improved: Upgrade to Electron 25. Note that it means that [Windows 7/8/8.1 are no longer supported](https://www.electronjs.org/blog/windows-7-to-8-1-deprecation-notice) (https://github.com/laurent22/joplin/pull/8451) (https://github.com/laurent22/joplin/issues/8258 by Henry Heino)
    Fixed: Fix issue with Apple Silicon version (https://github.com/laurent22/joplin/pull/8452) (https://github.com/laurent22/joplin/issues/6052 by Noah Nash)

Among the 3 changes, I suspect the second one the most, the electron update. It must have changed something for users who use IPv6. I'm on Linux, others have had the problem on Windows or MacOS. The operating system is therefore not the cause.

**Comment by laurent22** [comment]
Would you mind trying with the latest pre-release? As it includes an upgrade of Electron which may fix the issue:

https://github.com/laurent22/joplin/releases/tag/v2.13.5

**Comment by V1ck3s** [comment]
@laurent22 Hello, I tested with this version and the error is still there. I am putting back some of the logs.

```2023-11-10 20:13:26: JoplinServerApi: "Could not acquire session:", "undefined", "
", "FetchError: request to https://xxxxxxxxxxx.fr/api/sessions failed, reason: connect ECONNREFUSED 2b11:aea:ipv6XXXXXXXXX
Code: ECONNREFUSED
FetchError: request to https://xxxxxxxxxxx.fr/api/sessions failed, reason: connect ECONNREFUSED 2b11:aea:ipv6XXXXXXXXX
    at ClientRequest.<anonymous> (/tmp/.mount_JoplinKzPPky/resources/app.asar/node_modules/@joplin/lib/node_modules/node-fetch/lib/index.js:1491:11)
    at ClientRequest.emit (node:events:513:28)
    at TLSSocket.socketErrorListener (node:_http_client:502:9)
    at TLSSocket.emit (node:events:513:28)
    at emitErrorNT (node:internal/streams/destroy:151:8)
    at emitErrorCloseNT (node:internal/streams/destroy:116:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)"
2023-11-10 20:13:26: Synchronizer: "FetchError: request to https://xxxxxxxxxxx.fr/api/sessions failed, reason: connect ECONNREFUSED 2b11:aea:ipv6XXXXXXXXX
Code: ECONNREFUSED
FetchError: request to https://xxxxxxxxxxx.fr/api/sessions failed, reason: connect ECONNREFUSED 2b11:aea:ipv6XXXXXXXXX
    at ClientRequest.<anonymous> (/tmp/.mount_JoplinKzPPky/resources/app.asar/node_modules/@joplin/lib/node_modules/node-fetch/lib/index.js:1491:11)
    at ClientRequest.emit (node:events:513:28)
    at TLSSocket.socketErrorListener (node:_http_client:502:9)
    at TLSSocket.emit (node:events:513:28)
    at emitErrorNT (node:internal/streams/destroy:151:8)
    at emitErrorCloseNT (node:internal/streams/destroy:116:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)"```

**Comment by laurent22** [comment]
Only clue I could find is this: https://github.com/electron/electron/pull/35999#pullrequestreview-1271747380

But that's for Electron 18 and we've already been using it since v2.9.1.

The Electron 25 upgrade was this commit: https://github.com/laurent22/joplin/commit/a354feb626ee7d3f150b33cf64f281d181bb7132 which upgraded from v19 to v25, but not sure what could have been broken there.

**Comment by personalizedrefrigerator** [comment]
> Only clue I could find is this: https://github.com/electron/electron/pull/35999#pullrequestreview-1271747380
>
> But that's for Electron 18 and we've already been using it since v2.9.1.


The stack trace above mentions `node-fetch`. (This looks similar to [this issue](https://github.com/node-fetch/node-fetch/issues/1624#issuecomment-1407717012), which includes a suggested fix).

**Comment by laurent22** [comment]
Brilliant, thanks for looking into it @personalizedrefrigerator. I did check node-fetch at some point but didn't come across this issue

**Comment by personalizedrefrigerator** [comment]
It would be helpful if someone experiencing this issue could try the following (in a recent version of Joplin):
1. Open the development tools (Help > toggle development tools)
2. Run `require('dns').setDefaultResultOrder('ipv4first');` in the console (*Note*: This *might* need to be run soon after app startup)
    - [According to this issue](https://github.com/node-fetch/node-fetch/issues/1624#issuecomment-1407717012), this reverts an address resolution change made in Node 17, which might be causing this issue (or exposing an underlying issue related to how Joplin or a dependency handles ipv6).
3. Try to sync again

**Comment by V1ck3s** [comment]
I can try this tonight.

**Comment by V1ck3s** [comment]
@personalizedrefrigerator It works !

**Comment by romu70** [comment]
Same here, it works.

**Comment by mu88** [comment]
@laurent22 I tested this fix today with Joplin `2.12.19` on my Win11 laptop and running `require('dns').setDefaultResultOrder('ipv4first');` returns `undefined` but fixes the sync issue. However, when restarting Joplin, the issue is back again, meaning that I would have to run this command on every app start 🤔

**Comment by V1ck3s** [comment]
> @laurent22 I tested this fix today with Joplin `2.12.19` on my Win11 laptop and running `require('dns').setDefaultResultOrder('ipv4first');` returns `undefined` but fixes the sync issue. However, when restarting Joplin, the issue is back again, meaning that I would have to run this command on every app start 🤔

Already fixed, update your app.
